types/src/database_sql.rs

```diff
@@ -388,10 +390,68 @@ impl Database {
         todo!("Implement image annotations")
     }
 
-    pub async fn update_word(&self, word: &AnnotatedFormUpdate) -> Result<Uuid> {
-        query_file!("queries/update_word.sql", word.id, word.source.value())
-            .execute(&self.client)
-            .await?;
+    pub async fn update_word(&self, word: AnnotatedFormUpdate) -> Result<Uuid> {
+        let source = word.source.into_vec();
+        let commentary = word.commentary.into_vec();
+
+        query_file!(
+            "queries/update_word.sql",
+            word.id,
+            &source as _,
+            &commentary as _,
+        )
+        .execute(&self.client)
+        .await?;
+
+        // If word segmentation was not changed, then return early since SQL update queries need to be called.
+        if word.segments.is_undefined() {
+            return Ok(word.id);
+        }
+
+        let (doc_id, gloss, word_id, index, morpheme, role): (
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+            Vec<_>,
+        ) = word
+            .segments
+            .take()
+            .unwrap()
+            .into_iter()
+            .enumerate()
+            .map(move |(index, segment)| {
+                (
+                    word.doc_id,
+                    segment.gloss,
+                    word.id,
+                    index as i64, // index of the segment in the word
+                    segment.morpheme,
+                    segment.role,
+                )
+            })
+            .multiunzip();
+
+        query_file!(
+            "queries/upsert_local_morpheme_glosses.sql",
+            &*doc_id,
+            &*gloss
+        )
+        .execute(&self.client)
+        .await?;
+
+        query_file!(
+            "queries/upsert_many_word_segments.sql",
+            &*doc_id,
+            &*gloss,
```

Comment:
Okay so AFAICT you are using the glosses from the front-end straight-up with no conversion, and those glosses come from TAOC. That's a no-go. The TAOC tags are _not_ the same as the internal ones, though many/most of them appear the same. TAOC is just the main format where we currently have a 1:1 mapping between internal word parts and TAOC word parts (and the one where they're the most similar). So you still need to convert these glosses at some point.

I know it's confusing... Perhaps we can come up with a better scheme at some point.

Reply:
Maybe you're using the internal gloss already when you get the TAOC tag in the front-end, so it's all sorted?

Reply:
Yes, you're right in how I'm using the glosses. I'll see if I can convert those TAOC glosses into the internal ones. Would it make sense to do these in the front-end before sending it over into the back-end, or just have the back-end handle it?